File: nodejs/backend/src/middleware/logging.ts
import { Request, Response, NextFunction } from 'express';

export function logRequests(req: Request, res: Response, next: NextFunction): void {
  const start = process.hrtime.bigint();

  res.on('finish', () => {
    const elapsedMs = Number(process.hrtime.bigint() - start) / 1_000_000;
    console.log(
      JSON.stringify({
        level: 'info',
        method: req.method,
        path: req.path,
        status: res.statusCode,
        elapsedMs: Math.round(elapsedMs * 100) / 100,
        msg: 'request handled',
      }),
    );
  });

  next();
}

File: rust/backend/src/middleware/logging.rs
use axum::{extract::Request, middleware::Next, response::Response};
use std::time::Instant;

pub async fn log_requests(req: Request, next: Next) -> Response {
    let method = req.method().clone();
    let uri = req.uri().clone();
    let start = Instant::now();

    let response = next.run(req).await;

    tracing::info!(
        method = %method,
        path = %uri.path(),
        status = response.status().as_u16(),
        elapsed_ms = start.elapsed().as_millis(),
        "request handled"
    );

    response
}
